# iNaturalist status updates by state

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
State by state establish the changes that need to be made:
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive list
4. Attempt to match the state statuses to an IUCN equivalent


### 1. iNaturalist statuses

In [96]:
import sys
import os
projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
sys.path.append(os.path.abspath(projectdir + "source-code/includes"))
import list_functions as lf
import pandas as pd

sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
taxastatus.head(3)

id taxon_id  user_id place_id source_id  \
0  166449    38493  1138587     7830             
1  234788   918383   702203     9994             
2  234789   918383   702203     7308             

                            authority status  \
0  Flora and Fauna Guarantee Act 1988     CR   
1           Atlas of Living Australia    NT    
2           Atlas of Living Australia     LC   

                                                 url description geoprivacy  \
0                                                                  obscured   
1  https://bie.ala.org.au/species/https://id.biod...                          
2  https://bie.ala.org.au/species/https://id.biod...                          

   ...         genus specificEpithet infraspecificEpithet  \
0  ...     Eulamprus      kosciuskoi                        
1  ...  Chiloschista     phyllorhiza                        
2  ...  Chiloschista     phyllorhiza                        

               modified            scientificName taxonRank  \
0  2021-03-01T10:35:01Z      Eulamprus kosciuskoi   species   
1  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   
2  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   

                                          references preferred_common_name  \
0  http://reptile-database.reptarium.cz/search.ph...                         
1  http://www.catalogueoflife.org/annual-checklis...                         
2  http://www.catalogueoflife.org/annual-checklis...                         

  is_active current_synonymous_taxon_ids  
0                                         
1                                         
2                                         

[3 rows x 34 columns]

In [97]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])
inatstatuses = filter_state_statuses("Northern Territory|NT NRETAS", " ")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id user_id place_id source_id                  authority  \
1001    162724  1134239  702203     9994                   Northern Territory   
927     165448    12647  702203     9994                   Northern Territory   
289     263901  1289379  702203     9994            Atlas of Living Australia   
1251    152435    19250  708886     9994     16652                  NT NRETAS   
215     170163    20166  702203     9994            Atlas of Living Australia   
1249    152433    38633  708886     9994     16652                  NT NRETAS   
1252    152436    40743  708886     9994     16652                  NT NRETAS   
1248    152432    41326  708886     9994     16652                  NT NRETAS   
986     139906   698942             9994                                        
1247    152431    73180  708886     9994     16652                  NT NRETAS   
1250    152434    74425  708886     9994     16652                  NT NRETAS   
1       234788   918383  702203     9994            Atlas of Living Australia   

             status                                                url  \
1001             LC  https://bie.ala.org.au/species/https://id.biod...   
927              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
289              VU  https://bie.ala.org.au/species/urn:lsid:biodiv...   
1251     endangered  https://lists.ala.org.au/speciesListItem/list/...   
215              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
1249     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1252     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1248     endangered  https://lists.ala.org.au/speciesListItem/list/...   
986   least concern  https://bie.ala.org.au/species/http://id.biodi...   
1247     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1250     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1               NT   https://bie.ala.org.au/species/https://id.biod...   

                          description geoprivacy  ...         genus  \
1001                                        open  ...    Cryptandra   
927                                         open  ...    Epthianura   
289                                         open  ...      Chloebia   
1251                                    obscured  ...     Polytelis   
215                                               ...         Ninox   
1249                                    obscured  ...   Bellatorias   
1252                                    obscured  ...                 
1248                                    obscured  ...    Macroderma   
986   Atlas of Living Australia (ALA)             ...      Duboisia   
1247                                     private  ...     Pezoporus   
1250                                    obscured  ...  Hipposideros   
1                                                 ...  Chiloschista   

     specificEpithet infraspecificEpithet              modified  \
1001         gemmata                       2020-09-25T18:01:37Z   
927           crocea                       2021-09-17T08:46:17Z   
289         gouldiae                       2022-10-20T02:55:37Z   
1251      alexandrae                       2019-08-27T01:09:01Z   
215        connivens                       2021-07-28T02:17:03Z   
1249           obiri                       2019-04-30T15:19:21Z   
1252                                                              
1248           gigas                       2019-08-27T01:58:05Z   
986        hopwoodii                       2019-02-16T13:19:18Z   
1247    occidentalis                       2019-08-27T01:09:27Z   
1250       inornatus                       2019-01-05T10:25:06Z   
1        phyllorhiza                       2022-01-08T03:30:36Z   

                scientificName taxonRank  \
1001        Cryptandra gemmata   species   
927          Epthianura crocea   species   
289          Chloebia gouldiae   species   
1251      Polyte

### 2. iNaturalist taxonomy

In [98]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists
Sensitive list: `geoprivacy` = `obscured`
overrides

In [99]:
# %%script echo skipping # comment this line to download dataset from lists.ala.org.au the web and save locally
# Download lists data. Retrieve binomial and trinomial names from GBIF. Save locally to CSV

sensitivelist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr492?max=10000&includeKVP=true")
sensitivelist = lf.kvp_to_columns(sensitivelist)
sensitivelist.to_csv(sourcedir + "nt-ala-sensitive.csv", index=False)

conservationlist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr651?max=10000&includeKVP=true")
conservationlist = lf.kvp_to_columns(conservationlist)
conservationlist.to_csv(sourcedir + "nt-ala-conservation.csv", index=False)

In [100]:
# Read sensitive list data
sensitivelist = pd.read_csv(sourcedir + "nt-ala-sensitive.csv", dtype=str)
sensitivelist['status'] = 'sensitive'
sensitivelist['bionet_geoprivacy'] = 'obscured'
sensitivelist

id                    name               commonName  \
0  2058921        Macroderma gigas                Ghost Bat   
1  2058928   Hipposideros stenotis  Northern Leaf-nosed Bat   
2  2058925   Hipposideros inornata    Arnhem Leaf-nosed Bat   
3  2058920  Pezoporus occidentalis             Night Parrot   
4  2058926    Polytelis alexandrae       Alexandra's Parrot   
5  2058929        Falco hypoleucos              Grey Falcon   
6  2058922       Bellatorias obiri      Arnhem Land Egernia   
7  2058923           Attacus wardi               Atlas Moth   
8  2058924      Ogyris iphis doddi             Dodd’s Azure   
9  2058927      Candalides geminus                      NaN   

                  scientificName  \
0               Macroderma gigas   
1          Hipposideros stenotis   
2         Hipposideros inornatus   
3         Pezoporus occidentalis   
4           Polytelis alexandrae   
5  Falco (Hierofalco) hypoleucos   
6              Bellatorias obiri   
7                  Attacus wardi   
8             Ogyris iphis doddi   
9          Erina geminus geminus   

                                                lsid dataResourceUid  \
0  https://biodiversity.org.au/afd/taxa/63bc796a-...           dr492   
1  https://biodiversity.org.au/afd/taxa/26fe0f53-...           dr492   
2  https://biodiversity.org.au/afd/taxa/5d2dab40-...           dr492   
3  https://biodiversity.org.au/afd/taxa/c630f3b0-...           dr492   
4  https://biodiversity.org.au/afd/taxa/be7a08f5-...           dr492   
5  https://biodiversity.org.au/afd/taxa/4c73a934-...           dr492   
6  https://biodiversity.org.au/afd/taxa/2afc8501-...           dr492   
7  https://biodiversity.org.au/afd/taxa/8a05008e-...           dr492   
8  https://biodiversity.org.au/afd/taxa/ae3ab4c9-...           dr492   
9  https://biodiversity.org.au/afd/taxa/14d46baa-...           dr492   

                                           kvpValues          vernacular name  \
0  [{'key': 'vernacular name', 'value': 'Ghost Ba...                Ghost Bat   
1  [{'key': 'vernacular name', 'value': 'Northern...  Northern Leaf-nosed bat   
2  [{'key': 'vernacular name', 'value': 'Arnhem L...    Arnhem Leaf-nosed Bat   
3  [{'key': 'vernacular name', 'value': 'Night Pa...             Night Parrot   
4  [{'key': 'vernacular name', 'value': 'Princess...          Princess Parrot   
5  [{'key': 'vernacular name', 'value': 'Grey Fal...              Grey Falcon   
6  [{'key': 'vernacular name', 'value': 'Arnhemla...       Arnhemland Egernia   
7  [{'key': 'vernacular name', 'value': 'Atlas Mo...               Atlas Moth   
8  [{'key': 'vernacular name', 'value': 'Dodd’s A...             Dodd’s Azure   
9  [{'key': 'vernacular name', 'value': 'Twin Dus...          Twin Dusky-blue   

                     Criteria  \
0  cave roost related records   
1  cave roost related records   
2  cave roost related records   
3                         NaN   
4             nesting records   
5             nesting records   
6                         NaN   
7                         NaN   
8                         NaN   
9      eastern sub-population   

                                      Denature level generalisation  \
0  Round coordinate value to 0.05 decimal degrees...           10km   
1  Round coordinate value to 0.05 decimal degrees...           10km   
2  Round coordinate value to 0.05 decimal degrees...           10km   
3  Round coordinate value to 0.5 decimal degrees ...          100km   
4  Round coordinate value to 0.1 decimal degrees ...           10km   
5  Round coordinate value to 0.1 decimal degrees ...           10km   
6  Round coordinate value to 0.1 decimal degrees ...           10km   
7  Round coordinate value to 0.1 decimal degrees ...           10km   
8  Round coordinate value to 0.1 decimal degrees ...           10km   
9  Round coordinate value to 0.1 decimal degrees ...           10km   

      status bionet_geoprivacy  
0  sensitive          obscured  
1  sensitive        

In [101]:
# Read conservation list data
conservationlist = pd.read_csv(sourcedir + "nt-ala-conservation.csv", dtype=str)
conservationlist['bionet_geoprivacy'] = 'open'
conservationlist

id                           name                        commonName  \
0    4029368             Dasyuroides byrnei                            Kowari   
1    4029258              Phascogale calura             Red-tailed Phascogale   
2    4029309               Pseudomys fieldi                   Shark Bay Mouse   
3    4029399             Dasyurus geoffroii                     Western Quoll   
4    4029343                  Acacia latzii                     Latz's Wattle   
..       ...                            ...                               ...   
199  4029304                Leipoa ocellata                        Malleefowl   
200  4029285            Dasyurus hallucatus                    Northern Quoll   
201  4029371           Pedionomus torquatus                   Plains-wanderer   
202  4029227         Vincentrachia desmonda            Saddle Creek Rocksnail   
203  4029364  Amytornis modestus indulkanna  Thick-billed Grasswren (western)   

                                scientificName  \
0                           Dasyuroides byrnei   
1                            Phascogale calura   
2                             Pseudomys fieldi   
3                           Dasyurus geoffroii   
4                                Acacia latzii   
..                                         ...   
199                            Leipoa ocellata   
200                        Dasyurus hallucatus   
201                       Pedionomus torquatus   
202                     Vincentrachia desmonda   
203  Amytornis (Amytornis) modestus indulkanna   

                                                  lsid dataResourceUid  \
0    https://biodiversity.org.au/afd/taxa/c342ff42-...           dr651   
1    https://biodiversity.org.au/afd/taxa/36b436b1-...           dr651   
2    https://biodiversity.org.au/afd/taxa/edcf01fa-...           dr651   
3    https://biodiversity.org.au/afd/taxa/a2260672-...           dr651   
4     https://id.biodiversity.org.au/node/apni/2906346           dr651   
..                                                 ...             ...   
199  https://biodiversity.org.au/afd/taxa/c44c9098-...           dr651   
200  https://biodiversity.org.au/afd/taxa/5d7aeda8-...           dr651   
201  https://biodiversity.org.au/afd/taxa/30b4b2e5-...           dr651   
202  https://biodiversity.org.au/afd/taxa/2ff5b7ab-...           dr651   
203  https://biodiversity.org.au/afd/taxa/1309cbb2-...           dr651   

                                             kvpValues                 status  \
0    [{'key': 'status', 'value': 'Vulnerable'}, {'k...             Vulnerable   
1    [{'key': 'status', 'value': 'Vulnerable'}, {'k...             Vulnerable   
2    [{'key': 'status', 'value': 'Vulnerable'}, {'k...             Vulnerable   
3    [{'key': 'status', 'value': 'Vulnerable'}, {'k...             Vulnerable   
4    [{'key': 'family', 'value': 'Fabaceae'}, {'key...             Vulnerable   
..                                                 ...                    ...   
199  [{'key': 'status', 'value': 'Critically Endang...  Critically Endangered   
200  [{'key': 'status', 'value': 'Critically Endang...  Critically Endangered   
201  [{'key': 'status', 'value': 'Critically Endang...  Critically Endangered   
202  [{'key': 'status', 'value': 'Critically Endang...  Critically Endangered   
203  [{'key': 'status', 'value': 'Critically Endang...  Critically Endangered   

                   sourceStatus                   vernacular name  \
0    Vulnerable (extinct in NT)                            Kowari   
1    Vulnerable (extinct in NT)             Red-tailed phascogale   
2    Vulnerable (extinct in NT)                   Shark Bay mouse   
3    Vulnerable (extinct in NT)                     Western quoll   
4                    Vulnerable                     Latz's wattle   
..                          ...                               ...   
199       Critically Endangered                        Malleefowl   
200       Critically E

In [102]:
# join them in a way that works for inat (eg sensitive list, geoprivacy = 'obscured'
statelist = pd.concat([sensitivelist[['name', 'bionet_geoprivacy', 'status', 'lsid']],
                    conservationlist[['name', 'bionet_geoprivacy', 'status', 'lsid']]]).drop_duplicates()
statelist

name bionet_geoprivacy                 status  \
0                 Macroderma gigas          obscured              sensitive   
1            Hipposideros stenotis          obscured              sensitive   
2            Hipposideros inornata          obscured              sensitive   
3           Pezoporus occidentalis          obscured              sensitive   
4             Polytelis alexandrae          obscured              sensitive   
..                             ...               ...                    ...   
199                Leipoa ocellata              open  Critically Endangered   
200            Dasyurus hallucatus              open  Critically Endangered   
201           Pedionomus torquatus              open  Critically Endangered   
202         Vincentrachia desmonda              open  Critically Endangered   
203  Amytornis modestus indulkanna              open  Critically Endangered   

                                                  lsid  
0    https://biodiversity.org.au/afd/taxa/63bc796a-...  
1    https://biodiversity.org.au/afd/taxa/26fe0f53-...  
2    https://biodiversity.org.au/afd/taxa/5d2dab40-...  
3    https://biodiversity.org.au/afd/taxa/c630f3b0-...  
4    https://biodiversity.org.au/afd/taxa/be7a08f5-...  
..                                                 ...  
199  https://biodiversity.org.au/afd/taxa/c44c9098-...  
200  https://biodiversity.org.au/afd/taxa/5d7aeda8-...  
201  https://biodiversity.org.au/afd/taxa/30b4b2e5-...  
202  https://biodiversity.org.au/afd/taxa/2ff5b7ab-...  
203  https://biodiversity.org.au/afd/taxa/1309cbb2-...  

[214 rows x 4 columns]

In [103]:

# retrieve binomial and trinomial names from GBIF
parsednames = lf.gbifparse(statelist)
parsednames.to_csv(sourcedir + "nt-gbif.csv")
statelist = statelist.merge(parsednames[['scientificName','canonicalName']],how="left",left_on="name",right_on="scientificName").drop_duplicates()
statelist

name bionet_geoprivacy                 status  \
0                 Macroderma gigas          obscured              sensitive   
1            Hipposideros stenotis          obscured              sensitive   
2            Hipposideros inornata          obscured              sensitive   
3           Pezoporus occidentalis          obscured              sensitive   
5             Polytelis alexandrae          obscured              sensitive   
..                             ...               ...                    ...   
221                Leipoa ocellata              open  Critically Endangered   
222            Dasyurus hallucatus              open  Critically Endangered   
223           Pedionomus torquatus              open  Critically Endangered   
224         Vincentrachia desmonda              open  Critically Endangered   
225  Amytornis modestus indulkanna              open  Critically Endangered   

                                                  lsid  \
0    https://biodiversity.org.au/afd/taxa/63bc796a-...   
1    https://biodiversity.org.au/afd/taxa/26fe0f53-...   
2    https://biodiversity.org.au/afd/taxa/5d2dab40-...   
3    https://biodiversity.org.au/afd/taxa/c630f3b0-...   
5    https://biodiversity.org.au/afd/taxa/be7a08f5-...   
..                                                 ...   
221  https://biodiversity.org.au/afd/taxa/c44c9098-...   
222  https://biodiversity.org.au/afd/taxa/5d7aeda8-...   
223  https://biodiversity.org.au/afd/taxa/30b4b2e5-...   
224  https://biodiversity.org.au/afd/taxa/2ff5b7ab-...   
225  https://biodiversity.org.au/afd/taxa/1309cbb2-...   

                    scientificName                  canonicalName  
0                 Macroderma gigas               Macroderma gigas  
1            Hipposideros stenotis          Hipposideros stenotis  
2            Hipposideros inornata          Hipposideros inornata  
3           Pezoporus occidentalis         Pezoporus occidentalis  
5             Polytelis alexandrae           Polytelis alexandrae  
..                             ...                            ...  
221                Leipoa ocellata                Leipoa ocellata  
222            Dasyurus hallucatus            Dasyurus hallucatus  
223           Pedionomus torquatus           Pedionomus torquatus  
224         Vincentrachia desmonda         Vincentrachia desmonda  
225  Amytornis modestus indulkanna  Amytornis modestus indulkanna  

[214 rows x 6 columns]

In [104]:
statelist['scientificName'] = statelist['canonicalName']
statelist

name bionet_geoprivacy                 status  \
0                 Macroderma gigas          obscured              sensitive   
1            Hipposideros stenotis          obscured              sensitive   
2            Hipposideros inornata          obscured              sensitive   
3           Pezoporus occidentalis          obscured              sensitive   
5             Polytelis alexandrae          obscured              sensitive   
..                             ...               ...                    ...   
221                Leipoa ocellata              open  Critically Endangered   
222            Dasyurus hallucatus              open  Critically Endangered   
223           Pedionomus torquatus              open  Critically Endangered   
224         Vincentrachia desmonda              open  Critically Endangered   
225  Amytornis modestus indulkanna              open  Critically Endangered   

                                                  lsid  \
0    https://biodiversity.org.au/afd/taxa/63bc796a-...   
1    https://biodiversity.org.au/afd/taxa/26fe0f53-...   
2    https://biodiversity.org.au/afd/taxa/5d2dab40-...   
3    https://biodiversity.org.au/afd/taxa/c630f3b0-...   
5    https://biodiversity.org.au/afd/taxa/be7a08f5-...   
..                                                 ...   
221  https://biodiversity.org.au/afd/taxa/c44c9098-...   
222  https://biodiversity.org.au/afd/taxa/5d7aeda8-...   
223  https://biodiversity.org.au/afd/taxa/30b4b2e5-...   
224  https://biodiversity.org.au/afd/taxa/2ff5b7ab-...   
225  https://biodiversity.org.au/afd/taxa/1309cbb2-...   

                    scientificName                  canonicalName  
0                 Macroderma gigas               Macroderma gigas  
1            Hipposideros stenotis          Hipposideros stenotis  
2            Hipposideros inornata          Hipposideros inornata  
3           Pezoporus occidentalis         Pezoporus occidentalis  
5             Polytelis alexandrae           Polytelis alexandrae  
..                             ...                            ...  
221                Leipoa ocellata                Leipoa ocellata  
222            Dasyurus hallucatus            Dasyurus hallucatus  
223           Pedionomus torquatus           Pedionomus torquatus  
224         Vincentrachia desmonda         Vincentrachia desmonda  
225  Amytornis modestus indulkanna  Amytornis modestus indulkanna  

[214 rows x 6 columns]

In [105]:
statelist = pd.DataFrame(statelist[['scientificName','status','bionet_geoprivacy','lsid']])
statelist

scientificName                 status bionet_geoprivacy  \
0                 Macroderma gigas              sensitive          obscured   
1            Hipposideros stenotis              sensitive          obscured   
2            Hipposideros inornata              sensitive          obscured   
3           Pezoporus occidentalis              sensitive          obscured   
5             Polytelis alexandrae              sensitive          obscured   
..                             ...                    ...               ...   
221                Leipoa ocellata  Critically Endangered              open   
222            Dasyurus hallucatus  Critically Endangered              open   
223           Pedionomus torquatus  Critically Endangered              open   
224         Vincentrachia desmonda  Critically Endangered              open   
225  Amytornis modestus indulkanna  Critically Endangered              open   

                                                  lsid  
0    https://biodiversity.org.au/afd/taxa/63bc796a-...  
1    https://biodiversity.org.au/afd/taxa/26fe0f53-...  
2    https://biodiversity.org.au/afd/taxa/5d2dab40-...  
3    https://biodiversity.org.au/afd/taxa/c630f3b0-...  
5    https://biodiversity.org.au/afd/taxa/be7a08f5-...  
..                                                 ...  
221  https://biodiversity.org.au/afd/taxa/c44c9098-...  
222  https://biodiversity.org.au/afd/taxa/5d7aeda8-...  
223  https://biodiversity.org.au/afd/taxa/30b4b2e5-...  
224  https://biodiversity.org.au/afd/taxa/2ff5b7ab-...  
225  https://biodiversity.org.au/afd/taxa/1309cbb2-...  

[214 rows x 4 columns]

### 4. Equivalent IUCN statuses

In [106]:
# iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild', 'Extinct'}
statelist.groupby(['status'])['status'].count()


status
Critically Endangered     20
Endangered                52
Extinct                   11
Vulnerable               121
sensitive                 10
Name: status, dtype: int64

In [107]:
iucnStatusMappings = {
    'sensitive':'Vulnerable'
}

### 5. Determine best place ID to use

In [108]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()

place_id  place_name          place_display_name    
9994      Northern Territory  Northern Territory, AU    12
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State sensitive list on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [109]:
# join to see which lists already have a status in inaturalist based on scientificName
# mergedstatuses = statelist[['name','status','bionet_geoprivacy', 'lsid']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='name',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['scientificName'])
mergedstatuses = statelist[['scientificName', 'bionet_geoprivacy', 'lsid']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='scientificName',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['scientificName'])
mergedstatuses


scientificName bionet_geoprivacy  \
145  Abrodictyum obscurum              open   
195  Acacia equisetifolia              open   
14          Acacia latzii              open   
146          Acacia peuce              open   
15    Acacia praetermissa              open   
..                    ...               ...   
165    Xylopia monosperma              open   
65        Zeuxine oblonga              open   
68          Zyzomys maini              open   
176     Zyzomys palatalis              open   
201  Zyzomys pedunculatus              open   

                                                  lsid status_id taxon_id  \
145   https://id.biodiversity.org.au/node/apni/7402565       NaN      NaN   
195   https://id.biodiversity.org.au/node/apni/2890781       NaN      NaN   
14    https://id.biodiversity.org.au/node/apni/2906346       NaN      NaN   
146   https://id.biodiversity.org.au/node/apni/2906202       NaN      NaN   
15    https://id.biodiversity.org.au/node/apni/2894855       NaN      NaN   
..                                                 ...       ...      ...   
165   https://id.biodiversity.org.au/node/apni/2903202       NaN      NaN   
65   https://id.biodiversity.org.au/taxon/apni/5141...       NaN      NaN   
68   https://biodiversity.org.au/afd/taxa/3f638397-...       NaN      NaN   
176  https://biodiversity.org.au/afd/taxa/54aa72cc-...       NaN      NaN   
201  https://biodiversity.org.au/afd/taxa/33ab483d-...       NaN      NaN   

    user_id description iucn authority status geoprivacy place_id  \
145     NaN         NaN  NaN       NaN    NaN        NaN      NaN   
195     NaN         NaN  NaN       NaN    NaN        NaN      NaN   
14      NaN         NaN  NaN       NaN    NaN        NaN      NaN   
146     NaN         NaN  NaN       NaN    NaN        NaN      NaN   
15      NaN         NaN  NaN       NaN    NaN        NaN      NaN   
..      ...         ...  ...       ...    ...        ...      ...   
165     NaN         NaN  NaN       NaN    NaN        NaN      NaN   
65      NaN         NaN  NaN       NaN    NaN        NaN      NaN   
68      NaN         NaN  NaN       NaN    NaN        NaN      NaN   
176     NaN         NaN  NaN       NaN    NaN        NaN      NaN   
201     NaN         NaN  NaN       NaN    NaN        NaN      NaN   

    place_display_name  
145                NaN  
195                NaN  
14                 NaN  
146                NaN  
15                 NaN  
..                 ...  
165                NaN  
65                 NaN  
68                 NaN  
176                NaN  
201                NaN  

[214 rows x 13 columns]

In [110]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = " Territory Parks and Wildlife Conservation Act 1976"
mergedstatuses['new_description'] = "Listed as Threatened - refer to https://nt.gov.au/environment"
mergedstatuses['new_url'] = mergedstatuses['lsid']
mergedstatuses['new_geoprivacy'] = "obscured"
mergedstatuses['new_place_id'] = '9994'  # Northern Territory
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['status'].fillna('Threatened')
mergedstatuses

scientificName bionet_geoprivacy  \
145  Abrodictyum obscurum              open   
195  Acacia equisetifolia              open   
14          Acacia latzii              open   
146          Acacia peuce              open   
15    Acacia praetermissa              open   
..                    ...               ...   
165    Xylopia monosperma              open   
65        Zeuxine oblonga              open   
68          Zyzomys maini              open   
176     Zyzomys palatalis              open   
201  Zyzomys pedunculatus              open   

                                                  lsid status_id taxon_id  \
145   https://id.biodiversity.org.au/node/apni/7402565       NaN      NaN   
195   https://id.biodiversity.org.au/node/apni/2890781       NaN      NaN   
14    https://id.biodiversity.org.au/node/apni/2906346       NaN      NaN   
146   https://id.biodiversity.org.au/node/apni/2906202       NaN      NaN   
15    https://id.biodiversity.org.au/node/apni/2894855       NaN      NaN   
..                                                 ...       ...      ...   
165   https://id.biodiversity.org.au/node/apni/2903202       NaN      NaN   
65   https://id.biodiversity.org.au/taxon/apni/5141...       NaN      NaN   
68   https://biodiversity.org.au/afd/taxa/3f638397-...       NaN      NaN   
176  https://biodiversity.org.au/afd/taxa/54aa72cc-...       NaN      NaN   
201  https://biodiversity.org.au/afd/taxa/33ab483d-...       NaN      NaN   

    user_id description iucn authority status  ... place_id  \
145     NaN         NaN  NaN       NaN    NaN  ...      NaN   
195     NaN         NaN  NaN       NaN    NaN  ...      NaN   
14      NaN         NaN  NaN       NaN    NaN  ...      NaN   
146     NaN         NaN  NaN       NaN    NaN  ...      NaN   
15      NaN         NaN  NaN       NaN    NaN  ...      NaN   
..      ...         ...  ...       ...    ...  ...      ...   
165     NaN         NaN  NaN       NaN    NaN  ...      NaN   
65      NaN         NaN  NaN       NaN    NaN  ...      NaN   
68      NaN         NaN  NaN       NaN    NaN  ...      NaN   
176     NaN         NaN  NaN       NaN    NaN  ...      NaN   
201     NaN         NaN  NaN       NaN    NaN  ...      NaN   

    place_display_name                                      new_authority  \
145                NaN   Territory Parks and Wildlife Conservation Act...   
195                NaN   Territory Parks and Wildlife Conservation Act...   
14                 NaN   Territory Parks and Wildlife Conservation Act...   
146                NaN   Territory Parks and Wildlife Conservation Act...   
15                 NaN   Territory Parks and Wildlife Conservation Act...   
..                 ...                                                ...   
165                NaN   Territory Parks and Wildlife Conservation Act...   
65                 NaN   Territory Parks and Wildlife Conservation Act...   
68                 NaN   Territory Parks and Wildlife Conservation Act...   
176                NaN   Territory Parks and Wildlife Conservation Act...   
201                NaN   Territory Parks and Wildlife Conservation Act...   

                                       new_description  \
145  Listed as Threatened - refer to https://nt.gov...   
195  Listed as Threatened - refer to https://nt.gov...   
14   Listed as Threatened - refer to https://nt.gov...   
146  Listed as Threatened - refer to https://nt.gov...   
15   Listed as Threatened - refer to https://nt.gov...   
..                                                 ...   
165  Listed as Threatened - refer to https://nt.gov...   
65   Listed as Threatened - refer to https://nt.gov...   
68   Listed as Threatened - refer to https://nt.gov...   
176  Listed as Threatened - refer to https://nt.gov...   
201  Listed as Threatened - refer to https://nt.gov...   

                                               new_url new_geoprivacy  \
145   https://id.biodiversity.org.au/node/apni/7402565       obscured

## Updates

In [111]:
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('scientificName')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','scientificName','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'scientificName':'taxon_name',
                                  'status_id':'id'})
updates

action              taxon_name      id taxon_id      status  \
6    UPDATE       Bellatorias obiri  152433    38633  endangered   
169  UPDATE       Bellatorias obiri  152433    38633  endangered   
67   UPDATE  Hipposideros inornatus  152434    74425  endangered   
1    UPDATE   Hipposideros stenotis  152436    40743  endangered   
0    UPDATE        Macroderma gigas  152432    41326  endangered   
186  UPDATE  Pezoporus occidentalis  152431    73180  endangered   
3    UPDATE  Pezoporus occidentalis  152431    73180  endangered   
4    UPDATE    Polytelis alexandrae  152435    19250  endangered   
114  UPDATE    Polytelis alexandrae  152435    19250  endangered   

    iucn_equivalent                                          authority  \
6        Vulnerable   Territory Parks and Wildlife Conservation Act...   
169      Vulnerable   Territory Parks and Wildlife Conservation Act...   
67       Vulnerable   Territory Parks and Wildlife Conservation Act...   
1        Vulnerable   Territory Parks and Wildlife Conservation Act...   
0        Vulnerable   Territory Parks and Wildlife Conservation Act...   
186      Vulnerable   Territory Parks and Wildlife Conservation Act...   
3        Vulnerable   Territory Parks and Wildlife Conservation Act...   
4        Vulnerable   Territory Parks and Wildlife Conservation Act...   
114      Vulnerable   Territory Parks and Wildlife Conservation Act...   

                                                   url geoprivacy place_id  \
6    https://biodiversity.org.au/afd/taxa/2afc8501-...   obscured     9994   
169  https://biodiversity.org.au/afd/taxa/2afc8501-...   obscured     9994   
67   https://biodiversity.org.au/afd/taxa/5d2dab40-...   obscured     9994   
1    https://biodiversity.org.au/afd/taxa/26fe0f53-...   obscured     9994   
0    https://biodiversity.org.au/afd/taxa/63bc796a-...   obscured     9994   
186  https://biodiversity.org.au/afd/taxa/c630f3b0-...   obscured     9994   
3    https://biodiversity.org.au/afd/taxa/c630f3b0-...   obscured     9994   
4    https://biodiversity.org.au/afd/taxa/be7a08f5-...   obscured     9994   
114  https://biodiversity.org.au/afd/taxa/be7a08f5-...   obscured     9994   

      username                                        description  
6    peggydnew  Listed as Threatened - refer to https://nt.gov...  
169  peggydnew  Listed as Threatened - refer to https://nt.gov...  
67   peggydnew  Listed as Threatened - refer to https://nt.gov...  
1    peggydnew  Listed as Threatened - refer to https://nt.gov...  
0    peggydnew  Listed as Threatened - refer to https://nt.gov...  
186  peggydnew  Listed as Threatened - refer to https://nt.gov...  
3    peggydnew  Listed as Threatened - refer to https://nt.gov...  
4    peggydnew  Listed as Threatened - refer to https://nt.gov...  
114  peggydnew  Listed as Threatened - refer to https://nt.gov...

## No status in iNaturalist via straight scientificName match
The NT records that didn't match up to a status in iNaturalist

In [112]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="scientificName",right_on="scientificName")
noinatstatus

scientificName bionet_geoprivacy  \
0    Abrodictyum obscurum              open   
1    Acacia equisetifolia              open   
2           Acacia latzii              open   
3            Acacia peuce              open   
4     Acacia praetermissa              open   
..                    ...               ...   
200    Xylopia monosperma              open   
201       Zeuxine oblonga              open   
202         Zyzomys maini              open   
203     Zyzomys palatalis              open   
204  Zyzomys pedunculatus              open   

                                                  lsid status_id taxon_id  \
0     https://id.biodiversity.org.au/node/apni/7402565       NaN      NaN   
1     https://id.biodiversity.org.au/node/apni/2890781       NaN      NaN   
2     https://id.biodiversity.org.au/node/apni/2906346       NaN      NaN   
3     https://id.biodiversity.org.au/node/apni/2906202       NaN      NaN   
4     https://id.biodiversity.org.au/node/apni/2894855       NaN      NaN   
..                                                 ...       ...      ...   
200   https://id.biodiversity.org.au/node/apni/2903202       NaN      NaN   
201  https://id.biodiversity.org.au/taxon/apni/5141...       NaN      NaN   
202  https://biodiversity.org.au/afd/taxa/3f638397-...       NaN      NaN   
203  https://biodiversity.org.au/afd/taxa/54aa72cc-...       NaN      NaN   
204  https://biodiversity.org.au/afd/taxa/33ab483d-...       NaN      NaN   

    user_id description iucn authority status  ...        phylum  \
0       NaN         NaN  NaN       NaN    NaN  ...           NaN   
1       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
2       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
3       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
4       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
..      ...         ...  ...       ...    ...  ...           ...   
200     NaN         NaN  NaN       NaN    NaN  ...           NaN   
201     NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
202     NaN         NaN  NaN       NaN    NaN  ...      Chordata   
203     NaN         NaN  NaN       NaN    NaN  ...      Chordata   
204     NaN         NaN  NaN       NaN    NaN  ...      Chordata   

             class        order       family    genus specificEpithet  \
0              NaN          NaN          NaN      NaN             NaN   
1    Magnoliopsida      Fabales     Fabaceae   Acacia   equisetifolia   
2    Magnoliopsida      Fabales     Fabaceae   Acacia          latzii   
3    Magnoliopsida      Fabales     Fabaceae   Acacia           peuce   
4    Magnoliopsida      Fabales     Fabaceae   Acacia    praetermissa   
..             ...          ...          ...      ...             ...   
200            NaN          NaN          NaN      NaN             NaN   
201     Liliopsida  Asparagales  Orchidaceae  Zeuxine         oblonga   
202       Mammalia     Rodentia      Muridae  Zyzomys           maini   
203       Mammalia     Rodentia      Muridae  Zyzomys       palatalis   
204       Mammalia     Rodentia      Muridae  Zyzomys    pedunculatus   

    infraspecificEpithet              modified taxonRank  \
0                    NaN                   NaN       NaN   
1                    NaN  2022-04-06T22:05:25Z   species   
2                    NaN  2022-04-07T02:06:43Z   species   
3                    NaN  2022-04-06T23:48:13Z   species   
4                    NaN  2022-04-05T02:10:02Z   species   
..                   ...                   ...       ...   
200                  NaN                   NaN       NaN   
201                  NaN  2021-07-16T02:50:44Z   species   
202                  NaN  2019-08-27T01:49:16Z   species   
203                  NaN  2019-11-22T22:46:28Z   species   
204                  NaN  2019-11-22T22:46:29Z   species   

                                            references  
0                                                

In [113]:
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions

scientificName bionet_geoprivacy  \
1    Acacia equisetifolia              open   
2           Acacia latzii              open   
3            Acacia peuce              open   
4     Acacia praetermissa              open   
5       Acacia undoolyana              open   
..                    ...               ...   
198      Vidumelon wattii              open   
201       Zeuxine oblonga              open   
202         Zyzomys maini              open   
203     Zyzomys palatalis              open   
204  Zyzomys pedunculatus              open   

                                                  lsid status_id taxon_id  \
1     https://id.biodiversity.org.au/node/apni/2890781       NaN      NaN   
2     https://id.biodiversity.org.au/node/apni/2906346       NaN      NaN   
3     https://id.biodiversity.org.au/node/apni/2906202       NaN      NaN   
4     https://id.biodiversity.org.au/node/apni/2894855       NaN      NaN   
5     https://id.biodiversity.org.au/node/apni/2891259       NaN      NaN   
..                                                 ...       ...      ...   
198  https://biodiversity.org.au/afd/taxa/3aedec6e-...       NaN      NaN   
201  https://id.biodiversity.org.au/taxon/apni/5141...       NaN      NaN   
202  https://biodiversity.org.au/afd/taxa/3f638397-...       NaN      NaN   
203  https://biodiversity.org.au/afd/taxa/54aa72cc-...       NaN      NaN   
204  https://biodiversity.org.au/afd/taxa/33ab483d-...       NaN      NaN   

    user_id description iucn authority status  ...        phylum  \
1       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
2       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
3       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
4       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
5       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
..      ...         ...  ...       ...    ...  ...           ...   
198     NaN         NaN  NaN       NaN    NaN  ...      Mollusca   
201     NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
202     NaN         NaN  NaN       NaN    NaN  ...      Chordata   
203     NaN         NaN  NaN       NaN    NaN  ...      Chordata   
204     NaN         NaN  NaN       NaN    NaN  ...      Chordata   

             class            order       family      genus specificEpithet  \
1    Magnoliopsida          Fabales     Fabaceae     Acacia   equisetifolia   
2    Magnoliopsida          Fabales     Fabaceae     Acacia          latzii   
3    Magnoliopsida          Fabales     Fabaceae     Acacia           peuce   
4    Magnoliopsida          Fabales     Fabaceae     Acacia    praetermissa   
5    Magnoliopsida          Fabales     Fabaceae     Acacia      undoolyana   
..             ...              ...          ...        ...             ...   
198     Gastropoda  Stylommatophora   Camaenidae  Vidumelon          wattii   
201     Liliopsida      Asparagales  Orchidaceae    Zeuxine         oblonga   
202       Mammalia         Rodentia      Muridae    Zyzomys           maini   
203       Mammalia         Rodentia      Muridae    Zyzomys       palatalis   
204       Mammalia         Rodentia      Muridae    Zyzomys    pedunculatus   

    infraspecificEpithet              modified taxonRank  \
1                    NaN  2022-04-06T22:05:25Z   species   
2                    NaN  2022-04-07T02:06:43Z   species   
3                    NaN  2022-04-06T23:48:13Z   species   
4                    NaN  2022-04-05T02:10:02Z   species   
5                    NaN  2022-04-05T02:34:14Z   species   
..                   ...                   ...       ...   
198                  NaN  2021-10-29T15:37:54Z   species   
201                  NaN  2021-07-16T02:50:44Z   species   
202                  NaN  2019-08-27T01:49:16Z   species   
203                  NaN  2019-11-22T22:46:28Z   species   
204                  NaN  2019-11-22T22:46:29Z   species   

                                  

In [114]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['scientificName'])
additions['action'] = 'ADD'
additions = additions[['action','scientificName','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'scientificName':'taxon_name',
                                      'id':'taxon_id',
                                  'status_id':'id'})
additions

action            taxon_name   id taxon_id      status iucn_equivalent  \
1      ADD  Acacia equisetifolia  NaN  1253756  Threatened      Vulnerable   
2      ADD         Acacia latzii  NaN  1254327  Threatened      Vulnerable   
3      ADD          Acacia peuce  NaN   465191  Threatened      Vulnerable   
4      ADD   Acacia praetermissa  NaN  1254561  Threatened      Vulnerable   
5      ADD     Acacia undoolyana  NaN  1254884  Threatened      Vulnerable   
..     ...                   ...  ...      ...         ...             ...   
198    ADD      Vidumelon wattii  NaN   114966  Threatened      Vulnerable   
201    ADD       Zeuxine oblonga  NaN   369267  Threatened      Vulnerable   
202    ADD         Zyzomys maini  NaN    45377  Threatened      Vulnerable   
203    ADD     Zyzomys palatalis  NaN    75238  Threatened      Vulnerable   
204    ADD  Zyzomys pedunculatus  NaN    45374  Threatened      Vulnerable   

                                             authority  \
1     Territory Parks and Wildlife Conservation Act...   
2     Territory Parks and Wildlife Conservation Act...   
3     Territory Parks and Wildlife Conservation Act...   
4     Territory Parks and Wildlife Conservation Act...   
5     Territory Parks and Wildlife Conservation Act...   
..                                                 ...   
198   Territory Parks and Wildlife Conservation Act...   
201   Territory Parks and Wildlife Conservation Act...   
202   Territory Parks and Wildlife Conservation Act...   
203   Territory Parks and Wildlife Conservation Act...   
204   Territory Parks and Wildlife Conservation Act...   

                                                   url geoprivacy place_id  \
1     https://id.biodiversity.org.au/node/apni/2890781   obscured     9994   
2     https://id.biodiversity.org.au/node/apni/2906346   obscured     9994   
3     https://id.biodiversity.org.au/node/apni/2906202   obscured     9994   
4     https://id.biodiversity.org.au/node/apni/2894855   obscured     9994   
5     https://id.biodiversity.org.au/node/apni/2891259   obscured     9994   
..                                                 ...        ...      ...   
198  https://biodiversity.org.au/afd/taxa/3aedec6e-...   obscured     9994   
201  https://id.biodiversity.org.au/taxon/apni/5141...   obscured     9994   
202  https://biodiversity.org.au/afd/taxa/3f638397-...   obscured     9994   
203  https://biodiversity.org.au/afd/taxa/54aa72cc-...   obscured     9994   
204  https://biodiversity.org.au/afd/taxa/33ab483d-...   obscured     9994   

      username                                        description  
1    peggydnew  Listed as Threatened - refer to https://nt.gov...  
2    peggydnew  Listed as Threatened - refer to https://nt.gov...  
3    peggydnew  Listed as Threatened - refer to https://nt.gov...  
4    peggydnew  Listed as Threatened - refer to https://nt.gov...  
5    peggydnew  Listed as Threatened - refer to https://nt.gov...  
..         ...                                                ...  
198  peggydnew  Listed as Threatened - refer to https://nt.gov...  
201  peggydnew  Listed as Threatened - refer to https://nt.gov...  
202  peggydnew  Listed as Threatened - refer to https://nt.gov...  
203  peggydnew  Listed as Threatened - refer to https://nt.gov...  
204  peggydnew  Listed as Threatened - refer to https://nt.gov...  

[143 rows x 12 columns]

In [115]:
# write these to the file
pd.concat([updates,additions]).to_csv(sourcedir + "nt.csv", index=False)

In [116]:
# what didnt match to a taxon?
unknownToInat = noinatstatus[noinatstatus['id'].isna()]
unknownToInat

scientificName bionet_geoprivacy  \
0       Abrodictyum obscurum              open   
20        Babingtonia behrii              open   
21       Baumea arthrophylla              open   
23   Bettongia lesueur graii              open   
29     Burmannia sp.Bathurst              open   
..                       ...               ...   
186       Typhonium mirabile              open   
188    Typhonium sp.Sandover              open   
189        Typhonium taylori              open   
199   Vincentrachia desmonda              open   
200       Xylopia monosperma              open   

                                                  lsid status_id taxon_id  \
0     https://id.biodiversity.org.au/node/apni/7402565       NaN      NaN   
20   https://id.biodiversity.org.au/taxon/apni/5131...       NaN      NaN   
21    https://id.biodiversity.org.au/node/apni/2890775       NaN      NaN   
23   https://biodiversity.org.au/afd/taxa/7a019451-...       NaN      NaN   
29                                       ALA_DR651_139       NaN      NaN   
..                                                 ...       ...      ...   
186   https://id.biodiversity.org.au/node/apni/2900370       NaN      NaN   
188                                       ALA_DR651_53       NaN      NaN   
189   https://id.biodiversity.org.au/node/apni/2906352       NaN      NaN   
199  https://biodiversity.org.au/afd/taxa/2ff5b7ab-...       NaN      NaN   
200   https://id.biodiversity.org.au/node/apni/2903202       NaN      NaN   

    user_id description iucn authority status  ... phylum class order family  \
0       NaN         NaN  NaN       NaN    NaN  ...    NaN   NaN   NaN    NaN   
20      NaN         NaN  NaN       NaN    NaN  ...    NaN   NaN   NaN    NaN   
21      NaN         NaN  NaN       NaN    NaN  ...    NaN   NaN   NaN    NaN   
23      NaN         NaN  NaN       NaN    NaN  ...    NaN   NaN   NaN    NaN   
29      NaN         NaN  NaN       NaN    NaN  ...    NaN   NaN   NaN    NaN   
..      ...         ...  ...       ...    ...  ...    ...   ...   ...    ...   
186     NaN         NaN  NaN       NaN    NaN  ...    NaN   NaN   NaN    NaN   
188     NaN         NaN  NaN       NaN    NaN  ...    NaN   NaN   NaN    NaN   
189     NaN         NaN  NaN       NaN    NaN  ...    NaN   NaN   NaN    NaN   
199     NaN         NaN  NaN       NaN    NaN  ...    NaN   NaN   NaN    NaN   
200     NaN         NaN  NaN       NaN    NaN  ...    NaN   NaN   NaN    NaN   

    genus specificEpithet infraspecificEpithet modified taxonRank references  
0     NaN             NaN                  NaN      NaN       NaN        NaN  
20    NaN             NaN                  NaN      NaN       NaN        NaN  
21    NaN             NaN                  NaN      NaN       NaN        NaN  
23    NaN             NaN                  NaN      NaN       NaN        NaN  
29    NaN             NaN                  NaN      NaN       NaN        NaN  
..    ...             ...                  ...      ...       ...        ...  
186   NaN             NaN                  NaN      NaN       NaN        NaN  
188   NaN             NaN                  NaN      NaN       NaN        NaN  
189   NaN             NaN                  NaN      NaN       NaN        NaN  
199   NaN             NaN                  NaN      NaN       NaN        NaN  
200   NaN             NaN                  NaN      NaN       NaN        NaN  

[62 rows x 36 columns]

### are there any that need to be removed?
NT sensitive list count: 10
NT conservation list count: 204
NT inat statuses count: 12

updates to inat status: 9
additional inat status: 143
NT statuses we can't find a taxon match for in iNaturalist: 62
total:  (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over:  that may need checking

In [117]:
# Stats
numsensitive = len(sensitivelist.index)
numconservation = len(conservationlist.index)
numupdates  = len(updates.index)
numadditions  = len(additions.index)
numnoinatstatus = len(noinatstatus.index)
numunknownToInat = len(unknownToInat.index)
d = {'Sensitive': [numsensitive],
    'Conservation': [numconservation],
    'Updates': [numupdates],
    'Additions': [numadditions],
    'No Inat Status': [numnoinatstatus],
    'Unknown to Inat': [numunknownToInat]}
statsdf = pd.DataFrame(data=d)
statsdf

Sensitive  Conservation  Updates  Additions  No Inat Status  \
0         10           204        9        143             205   

   Unknown to Inat  
0               62

In [118]:
# inat statuses that aren't in added or updated
inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]


status_id taxon_id user_id place_id source_id                  authority  \
1001    162724  1134239  702203     9994                   Northern Territory   
927     165448    12647  702203     9994                   Northern Territory   
289     263901  1289379  702203     9994            Atlas of Living Australia   
215     170163    20166  702203     9994            Atlas of Living Australia   
986     139906   698942             9994                                        
1       234788   918383  702203     9994            Atlas of Living Australia   

             status                                                url  \
1001             LC  https://bie.ala.org.au/species/https://id.biod...   
927              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
289              VU  https://bie.ala.org.au/species/urn:lsid:biodiv...   
215              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
986   least concern  https://bie.ala.org.au/species/http://id.biodi...   
1               NT   https://bie.ala.org.au/species/https://id.biod...   

                          description geoprivacy  ...         genus  \
1001                                        open  ...    Cryptandra   
927                                         open  ...    Epthianura   
289                                         open  ...      Chloebia   
215                                               ...         Ninox   
986   Atlas of Living Australia (ALA)             ...      Duboisia   
1                                                 ...  Chiloschista   

     specificEpithet infraspecificEpithet              modified  \
1001         gemmata                       2020-09-25T18:01:37Z   
927           crocea                       2021-09-17T08:46:17Z   
289         gouldiae                       2022-10-20T02:55:37Z   
215        connivens                       2021-07-28T02:17:03Z   
986        hopwoodii                       2019-02-16T13:19:18Z   
1        phyllorhiza                       2022-01-08T03:30:36Z   

                scientificName taxonRank  \
1001        Cryptandra gemmata   species   
927          Epthianura crocea   species   
289          Chloebia gouldiae   species   
215            Ninox connivens   species   
986         Duboisia hopwoodii   species   
1     Chiloschista phyllorhiza   species   

                                             references preferred_common_name  \
1001  http://plantsoftheworldonline.org/taxon/urn:ls...                         
927   http://www.birdlife.org/datazone/speciesfactsh...                         
289   https://www.birds.cornell.edu/clementschecklis...                         
215   http://www.birdlife.org/datazone/speciesfactsh...                         
986                             https://www.ala.org.au/                         
1     http://www.catalogueoflife.org/annual-checklis...                         

     is_active current_synonymous_taxon_ids  
1001                                         
927                                          
289                                          
215                                          
986                                          
1                                            

[6 rows x 34 columns]